scalasql/query/src-3/TableMacro.scala

```diff
@@ -0,0 +1,177 @@
+
+        ${
+          Expr.ofList(constructorValueParams.map { param =>
+            val paramTpe = paramType(param)
+            val tpe = subParam(paramTpe, TypeRepr.of[Sc])
+            val tpe2 = subParam(paramTpe, TypeRepr.of[SqlExpr])
+            (tpe.asType, tpe2.asType) match {
+              case ('[t], '[t2]) => '{ summonInline[Queryable.Row[t2, t]] }
+            }
+          })
+        }.apply(n)
+      }
+    }
+
+    val walkLabels0 = '{ () =>
+      ${
+        Expr.ofList(constructorValueParams.map { param =>
+          if (isTypeParamType(param))
+            paramType(param) match {
+              case AppliedType(tpeCtor, _) =>
+                tpeCtor.asType match {
+                  case '[
+                      type t[_[_]]; t] =>
+                    '{ summonInline[Table.ImplicitMetadata[t]].value.walkLabels0() }
+                }
+            }
+          else '{ Seq(${ Expr(param.name) }) }
+        })
+      }.flatten
+    }
+
+    val queryable = '{
+      (
+          walkLabels0: () => Seq[String],
+          dialect: DialectTypeMappers,
+          queryable: Table.Metadata.QueryableProxy
+      ) =>
+        new Table.Internal.TableQueryable(
+          walkLabels0,
+          (table: V[SqlExpr]) =>
+            ${
+              Expr.ofList(constructorValueParams.zipWithIndex.map { case (param, i) =>
+                val paramTpe = paramType(param)
+                val tpe = subParam(paramTpe, TypeRepr.of[Sc])
+                val tpe2 = subParam(paramTpe, TypeRepr.of[SqlExpr])
+                (tpe.asType, tpe2.asType) match {
+                  case ('[t], '[t2]) =>
+                    '{
+                      queryable[t2, t](${ Expr(i) }).walkExprs(
+                        ${ Select.unique('table.asTerm, param.name).asExprOf[t2] }
+                      )
+                    }
+                }
+              })
+            }.flatten,
+          construct0 = (args: Queryable.ResultSetIterator) =>
+            ${
+              Apply(
+                TypeApply(
+                  Select.unique(New(TypeIdent(TypeRepr.of[V].typeSymbol)), "<init>"),
+                  List(TypeTree.of[Sc])
+                ),
+                constructorValueParams.zipWithIndex.map { case (param, i) =>
+                  val paramTpe = paramType(param)
+                  val tpe = subParam(paramTpe, TypeRepr.of[Sc]).asType
+                  val tpe2 = subParam(paramTpe, TypeRepr.of[SqlExpr]).asType
+
+                  (tpe, tpe2) match {
+                    case ('[t], '[t2]) => '{ queryable[t2, t](${ Expr(i) }).construct(args) }.asTerm
+                  }
+                }
+              ).asExprOf[V[Sc]]
+            },
+          deconstruct0 = (r: V[Sc]) =>
+            ${
+              Apply(
+                TypeApply(
+                  Select.unique(New(TypeIdent(TypeRepr.of[V].typeSymbol)), "<init>"),
+                  List(TypeTree.of[SqlExpr])
+                ),
```

Comment:
Could this AST-constructions of the `new Foo` call be moved into a named helper method? To help avoid duplication, and to give it a name so it's obvious what it is without having to read through the AST structure

Reply:
Addressed this along with the deduplication of the `paramTpe`, `tpe`, and `tpe2` logic